pkg/storage/sqlcommon/sqlcommon.go

```diff
@@ -187,10 +190,74 @@ func NewSQLTupleIterator(rows *sql.Rows) *SQLTupleIterator {
 		rows:     rows,
 		resultCh: make(chan *storage.TupleRecord, 1),
 		errCh:    make(chan error, 1),
+		firstRow: nil,
+		mu:       sync.Mutex{},
 	}
 }
 
 func (t *SQLTupleIterator) next() (*storage.TupleRecord, error) {
+	t.mu.Lock()
+
+	if t.firstRow != nil {
```

Comment:
I don't understand why this check is here, if you call `head()` it will read one item and return it. If then you call `next()` it will return that same item, but i'd expect it to perform a read no matter what.

Reply:
No, calling `next()` does not depend on `head()`.  However, if we had called `head()`, we had already moved the internal iterator (because we actually called `t.rows.Next()`).  Thus, we need to consume this "firstRow" instead of calling `f.rows.Next()`.  Otherwise, that first item is lost.  I will update in the code comment to explain in more details.

Reply:
It seems like this logic could be simplified, and the interface more aligned with Go's `sql.Rows` interface, by making `Head` dependent on an initial call to `Next`. In Go's `sql.Rows` interface, a call to `Scan` must be preceded by a call to `Next`. From my perspective, it is least surprising that `Head` would function much the same.

Reply:
I did not choose to use the sql format because IMO it may introduce some weirdness. The current semantic Next() itself will return the first item and increment to the iterator.  The first item is lost and it is the caller's responsibility to store it. If we choose to follow the Scan/Next semantic, this may change how caller needs to manage their "Next" usage.